### PR TITLE
`tox -e update_docker_platforms`

### PR DESCRIPTION
**Pull request by mkoeppe**
Furthering my hidden agenda of bringing portability tools to the masses.

- We add `.devcontainer/portability-*/devcontainer.json` files for all platforms tested in our CI. This makes them selectable in GitHub Codespaces. 
- The files are generated by the new command `tox -e update_docker_platforms`, which also updates the workflow file `.github/workflows/docker.yml`.
- The command also [adds a table to the portability testing section in the developer's guide](https://deploy-preview-36954--sagemath.netlify.app/html/en/developer/portability_testing#using-our-pre-built-docker-images-published-on-ghcr-io) with links to GitHub Packages and friendly "Run in GitHub Codespaces" buttons.
- Because 'tis the season of Spielzeugneid, [we document how to run the portability testing in GitHub Codespaces via Docker-in-Docker](https://deploy-preview-36954--sagemath.netlify.app/html/en/developer/portability_testing#testing-sage-on-a-different-platform-using-docker), useful for those who cannot run Docker on their own computers. This uses a new devcontainer config "tox-docker-in-docker". 
- We make various other updates to the portability testing section in the developer's guide.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

- Depends on #37030 (split out from here)

### CODE REVIEW

**Comment by kwankyu** [comment]
This looks better

Screen Shot 2024-01-09 at 4 40 34 PM

```
.. list-table::
   :width: 200
   :widths: 50 20 50 50
   :header-rows: 1
   :stub-columns: 0

   * - Platform
     - System packages
     - Images
     -
   * - **ubuntu**-trusty-toolchain-gcc_9
     - minimal
     - |image-ubuntu-trusty-toolchain-gcc_9-minimal-with-system-packages| |image-ubuntu-trusty-toolchain-gcc_9-minimal-with-targets-pre| |image-ubuntu-trusty-toolchain-gcc_9-minimal-with-targets| |image-ubuntu-trusty-toolchain-gcc_9-minimal-with-targets-optional|
     - |codespace-ubuntu-trusty-toolchain-gcc_9-minimal|
   * -
     - standard
     - |image-ubuntu-trusty-toolchain-gcc_9-standard-with-system-packages| |image-ubuntu-trusty-toolchain-gcc_9-standard-with-targets-pre| |image-ubuntu-trusty-toolchain-gcc_9-standard-with-targets| |image-ubuntu-trusty-toolchain-gcc_9-standard-with-targets-optional|
     - |codespace-ubuntu-trusty-toolchain-gcc_9-standard|
   * -
     - maximal
     - |image-ubuntu-trusty-toolchain-gcc_9-maximal-with-system-packages| |image-ubuntu-trusty-toolchain-gcc_9-maximal-with-targets-pre|
     -  
```

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
Please don't commit the auto-generated devcontainer files. The would show up at https://github.com/codespaces/new?hide_repo_select=true&ref=develop&repo=597660615&skip_quickstart=true, and a drop down box should not contain more than 10-20 options to be useful. Also, if I'm not mistaken, the docs already explain how to start a devcontainer for the portability tests.

**Comment by kwankyu** [comment]
> Please don't commit the auto-generated devcontainer files. The would show up at https://github.com/codespaces/new?hide_repo_select=true&ref=develop&repo=597660615&skip_quickstart=true, and a drop down box should not contain more than 10-20 options to be useful. 

I experimented on it. The dev container files appear sorted alphabetically by their filenames, and the new dev container files (that start with "portability-") appear to the bottom of the dropdown box. Hence there is no inconvenience incurred by this PR. The conda dev container (perhaps your favorite) is still placed at the top. 

Screenshot 2024-01-09 at 8 14 19 PM

**Comment by tobiasdiez** [comment]
Thanks for testing, but I was already aware of the sorting and took this into account.

I would be fine with adding a few common configs, but not almost hundred.

**Comment by mkoeppe** [comment]
> the docs already explain how to start a devcontainer for the portability tests.

That's right. This PR lowers the barrier further -- in particular by making these options available in the devcontainer dropdown menu on PRs.

> I would be fine with adding a few common configs

There's no such thing, as I explained in detail in #36726.

**Comment by mkoeppe** [comment]
> This looks better
> 
> ```
> .. list-table::
>    :width: 200
>    :widths: 50 20 50 50
>    :header-rows: 1
>    :stub-columns: 0
> ```

Not sure about this. Have you tried this in a narrow window?

**Comment by kwankyu** [comment]
> Not sure about this. Have you tried this in a narrow window?

I just did. In the narrowest window (perhaps corresponding to the vertical mode of hand-held devices), the PR version looks okay. In this case, the font is too small in my version. Except this extreme case, my version shows always the four images stacked nicely and codespaces button separately on the right, which look tidy overall.

File: src/doc/en/developer/portability_testing.rst [hunk 1]
```diff
@@ -60,13 +60,33 @@ Make sure that your Docker client is configured to provide enough RAM
...
+
+   .. list-table::
+      :widths: 25 60
+      :header-rows: 0
+
+      * - `tox-docker-in-docker
+          <https://github.com/sagemath/sage/tree/develop/.devcontainer/tox-docker-in-docker>`_
+        - |codespace-tox-docker-in-docker|
+
 All examples in this section were obtained using Docker Desktop for
```
Comment (kwankyu) [inline src/doc/en/developer/portability_testing.rst]: Lines 63--80 may be put into `.. note::` as this paragraph breaks the normal flow of narration.

It seems that the list-table is not necessary. The link to the file is already in line 66--67. How about just displaying the codespaces button (this button would be part of the `.. note::`)?

**Comment by kwankyu** [comment]
Instead of `tox -e update_docker_platforms`, how about
```
tox -e update_devcontainer_files
```
as this is what is done? Just a mild suggestion.

**Comment by tobiasdiez** [comment]
> > I would be fine with adding a few common configs
> 
> There's no such thing, as I explained in detail in #36726.

There is, and for that system we already added a devcontainer file.

**Comment by tobiasdiez** [comment]
And please move the script out of tox. There doesn't seem anything specific in there that needs tox/an isolated environment. Perhaps we should introduce a new `tools` directory as a common place for development tools.

**Comment by mkoeppe** [comment]
> > > I would be fine with adding a few common configs
> > 
> > There's no such thing, as I explained in detail in #36726.
> 
> There is, and for that system we already added a devcontainer file.

Tobias, stop writing for an audience. You and I both know that you have no involvement whatsoever in maintaining the Sage distribution and the associated portability work. Such declarations from you about this effort just have no value in this discussion.

**Comment by mkoeppe** [comment]
> And please move the script out of tox.

Nope, that's where it belongs, it's the central place where the different system configurations are defined.

**Comment by mkoeppe** [comment]
> Instead of `tox -e update_docker_platforms`, how about
> 
> ```
> tox -e update_devcontainer_files
> ```
> 
> as this is what is done? Just a mild suggestion.

That's one thing that it does, but it also updates the file `.github/workflows/docker.yml`, which defines the CI Linux workflow.

**Comment by kwankyu** [comment]
> That's one thing that it does, but it also updates the file `.github/workflows/docker.yml`, which defines the CI Linux workflow.

I see. Then this
```
commands =
    bash -c 'DEFAULT_SYSTEM_FACTORS=$(echo \
                     ubuntu-{trusty-toolchain-gcc_9,xenial-toolchain-gcc_9,bionic-gcc_8,focal,jammy,lunar,mantic} \
                     debian-{buster-gcc_spkg,bullseye,bookworm,trixie,sid} \
                     linuxmint-{20.1,20.2,20.3,21,21.1,21.2} \
                     fedora-{30,31,32,33,34,35,36,37,38,39} \
                     centos-7-devtoolset-gcc_11 \
                     centos-stream-{8,9}-python3.9 \
                     almalinux-{8-python3.9,9-python3.11} \
                     gentoo-python{3.10,3.11} \
                     archlinux-latest \
                     opensuse-{15.3-gcc_11-python3.9,15.4-gcc_11-python3.10,15.5-gcc_11-python3.11} \
                     opensuse-tumbleweed{-python3.10,} \
                     conda-forge-python3.11 \
                     ubuntu-bionic-gcc_8-i386 \
                     debian-bullseye-i386); \    
``` 
is responsible for defining all linux platforms that our CI machine works on and thus sage officially supports. Right? 

Then it seems that this important list is buried deep in `tox.ini`. Would it be possible to pull this definition out of the quoted string to a top-level visible place (such as `setenv=` or something)? Even the color codes `COLOR_...` are more visible now.

**Comment by mkoeppe** [comment]
> Would it be possible to pull this definition out of the quoted string to a top-level visible place (such as `setenv=` or something)?

That's a good idea. I'll work on that

**Comment by mkoeppe** [comment]
But first here's another layout variant....

Screenshot 2024-01-09 at 9 11 01 PM

**Comment by kwankyu** [comment]
OK for me.

**Comment by tobiasdiez** [comment]
> > > > I would be fine with adding a few common configs
> > > 
> > > 
> > > There's no such thing, as I explained in detail in #36726.
> > 
> > 
> > There is, and for that system we already added a devcontainer file.
> 
> Tobias, stop writing for an audience. You and I both know that you have no involvement whatsoever in maintaining the Sage distribution and the associated portability work. Such declarations from you about this effort just have no value in this discussion.

Please stop these ad hominem attacks against reviewers. Moreover, it is surely against our code of conduct to try to limit the discussion of certain topics to a subgroup of the community (and even more so if a single person decides who is in that subgroup, or if it is based on who contributed to a certain part of the project before).

(Finally, I did actually have quite a few commits changing things in the sage-the-distro folders `build` and `pkgs`.)

**Comment by mkoeppe** [comment]
> > Would it be possible to pull this definition out of the quoted string to a top-level visible place (such as `setenv=` or something)?
> 
> That's a good idea. I'll work on that

Done in e663e2d

**Comment by kwankyu** [comment]
Looking at the rendered table with 3 columns, it seems that the codespaces button occupies too much space, and we still have a room for the column for "system packages". Thus I still prefer the 4 column format as suggested above. But I won't insist, as this seems to be a personal preference.

By the way, here

Screen Shot 2024-01-10 at 3 43 08 PM

and here

Screen Shot 2024-01-10 at 3 43 25 PM

there are broken links. Could they be fixed?

**Comment by mkoeppe** [comment]
> Looking at the rendered table with 3 columns, it seems that the codespaces button occupies too much space

I've adjusted the column width and another detail, please take a look.

**Comment by mkoeppe** [comment]
> there are broken links. Could they be fixed?

The one for conda was systematic, I'm suppressing this row now. The other one must have been some query rate limit in the badge generation service; this should go away if you clear the cache

**Comment by kwankyu** [comment]
> > Looking at the rendered table with 3 columns, it seems that the codespaces button occupies too much space
> 
> I've adjusted the column width and another detail, please take a look.

Looks good to me. Thanks.

**Comment by kwankyu** [comment]
> > there are broken links. Could they be fixed?
> 
> The one for conda was systematic, I'm suppressing this row now. The other one must have been some query rate limit in the badge generation service; this should go away if you clear the cache

I see no broken links now. Thanks.

File: src/doc/en/developer/portability_testing.rst (hunk 1 above)
Comment (kwankyu) [reply src/doc/en/developer/portability_testing.rst]: I am happy with this PR except this issue.

Comment (kwankyu) [reply src/doc/en/developer/portability_testing.rst]: Like this
```
Make sure that your Docker client is configured to provide enough RAM
to the containers (8 GB are a good choice). In Docker Desktop this
setting is in Preferences -> Resources -> Advanced.

.. NOTE::

   All examples in this section were obtained using Docker Desktop for
   Mac; but the `command-line user interface
   <https://docs.docker.com/engine/reference/commandline/cli/>`_ for the
   other platforms is identical.

   As an alternative, you can also run Docker in GitHub Codespaces
   (or another cloud service) using a container with the Docker-in-Docker
   feature. Sage provides a suitable dev container configuration
   `.devcontainer/tox-docker-in-docker
   <https://github.com/sagemath/sage/tree/develop/.devcontainer/tox-docker-in-docker>`_:

   .. ONLY:: html

      .. image:: https://github.com/codespaces/badge.svg
         :align: center
         :target: https://codespaces.new/sagemath/sage?devcontainer_path=.devcontainer%2Ftox-docker-in-docker%2Fdevcontainer.json

All major Linux distributions provide ready-to-use Docker images,   
```

Comment (kwankyu) [reply src/doc/en/developer/portability_testing.rst]: Screen Shot 2024-01-11 at 1 59 11 PM

Comment (mkoeppe) [reply src/doc/en/developer/portability_testing.rst]: Thanks, that's a nice improvement. I've committed it as 390c949db7b

**Review by kwankyu — APPROVED** [review]
LGTM.

**Comment by tobiasdiez** [comment]
> > And please move the script out of tox.
> 
> Nope, that's where it belongs, it's the central place where the different system configurations are defined.

That doesn't explain why one need to install tox to generate a new version of the devcontainer or update the docker yml file.

**Comment by tobiasdiez** [comment]
> Tox is a standard package.

But it's not a bootstrap dependency, while bash and python are...

**Comment by mkoeppe** [comment]
As you know, this is irrelevant because it is not running at bootstrap time. It is run by developers who then commit the changes to the generated files to git.

**Comment by tobiasdiez** [comment]
All current scripts that update/generate code work with the bootstrap dependencies, i.e. outside of sage's venv. I don't see a reason why we should depart from this established pattern.

**Comment by mkoeppe** [comment]
That's neither true (note `sage -fixdistributions`) nor by design, it's just a random observation.

**Comment by mkoeppe** [comment]
Let's get this in.

**Comment by kwankyu** [comment]
Enough lingering here.

**Pull request merged**